### PR TITLE
Composed Spec Execution with Streaming Output

### PR DESCRIPTION
**Pull request by swoogles**
I'm going to continue polishing this tomorrow as time allows, and it will need a few followup PRs but it's potentially mergeable now.

`ZIOSpecSpec` is still flagged as ignored for now.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -19,7 +19,8 @@ inThisBuild(
         "john@degoes.net",
         url("http://degoes.net")
       )
-    )
+    ),
+    Test / logBuffered := false // Enables streaming test output from composed specs
```
Comment (swoogles) [inline build.sbt:23]: We probably want to add this to our "Recommended SBT settings" in some form

File: test-sbt/jvm/src/main/scala/zio/test/sbt/ZTestRunner.scala [hunk 2]
```diff
@@ -141,30 +141,19 @@ class ZTestTaskPolicyDefaultImpl extends ZTestTaskPolicy {
...
+                  (
+                    MergedSpec(
+                      new ZTestTaskNew(
+                        existingNewTestTask.spec.taskDef,
+                        existingNewTestTask.spec.testClassLoader,
+                        existingNewTestTask.spec.sendSummary,
+                        existingNewTestTask.spec.args,
+                        existingNewTestTask.spec.newSpec <> taskNew.newSpec
+                      ),
+                      existingNewTestTask.merges + 1
```
Comment (adamgfraser) [inline test-sbt/jvm/src/main/scala/zio/test/sbt/ZTestRunner.scala:153]: Do we still need to keep track of the number of merges?

File: test-sbt/js/src/main/scala/zio/test/sbt/ZTestRunner.scala [hunk 3]
```diff
@@ -98,34 +98,24 @@ sealed class ZTestTask(
...
-                          testLoggers +!+ fullLayer
-                        )
-              events = ZTestEvent.from(spec, taskDef.fullyQualifiedName(), taskDef.fingerprint())
-              _     <- ZIO.foreach(events)(e => ZIO.attempt(eventHandler.handle(e)))
+              _ <- zioSpec
+                     .runSpec(FilteredSpec(zioSpec.spec, args), args)
+                     .provideLayer(
+                       fullLayer
+                     )
+              // TODO Confirm if/how these events needs to be handled
```
Comment (adamgfraser) [inline test-sbt/js/src/main/scala/zio/test/sbt/ZTestRunner.scala:115]: Probably better to open an issue for this.

File: test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala [hunk 4]
```diff
@@ -85,18 +99,14 @@ object ZTestFrameworkSpec {
...
-          s"${reset("info: ")}  ${Console.BLUE}Hello,",
+          s"${reset("info: ")}  ${red("- multi-line test")}",
+          s"${reset("info: ")}    ${Console.BLUE}Hello,",
           s"${reset("info: ")}${blue("World!")} did not satisfy ${cyan("equalTo(Hello, World!)")}",
-          s"${reset("info: ")}  ${assertSourceLocation()}",
-          s"${reset("info: ")}"
+          s"${reset("info: ")}    ${assertSourceLocation()}",
+          s"""${reset("info: ")}${ConsoleRenderer.render(Summary(0, 1, 0, ""))}"""
         ).mkString("\n")
-//          .mkString("\n")
```
Comment (adamgfraser) [inline test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala:96]: Dead code?

File: test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala [hunk 5]
```diff
@@ -18,64 +20,76 @@ object ZTestFrameworkSpec {
...
+        )
+      )
+      assertContains(
+        "logs ignored message",
+        messages,
+        Seq(
+          s"${reset("info:")}     ${yellow("-")} ${yellow("ignored test")}"
+        )
+      )
+    // We can't do this assertion anymore with the streaming approach
```
Comment (adamgfraser) [inline test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala:90]: Delete?

File: test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala [hunk 6]
```diff
@@ -18,64 +20,76 @@ object ZTestFrameworkSpec {
...
-    test("should report events")(testReportEvents()),
-    test("should report durations")(testReportDurations()),
-    test("should log messages")(testLogMessages()),
-    test("should correctly display colorized output for multi-line strings")(testColored()),
-    test("should test only selected test")(testTestSelection()),
-    test("should return summary when done")(testSummary()),
-    test("should use a shared layer without re-initializing it")(testSharedLayer()),
-    test("should warn when no tests are executed")(testNoTestsExecutedWarning())
+    // TODO restore once we are calculating durations again
+    //test("should report durations")(testReportDurations()),
```
Comment (adamgfraser) [inline test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala:24]: Could also be part of issue to track follow up items.

File: test-sbt/shared/src/main/scala/zio/test/sbt/BaseTestTask.scala [hunk 7]
```diff
@@ -39,82 +25,68 @@ abstract class BaseTestTask(
...
-    spec: ZIOSpecAbstract,
-    loggers: Array[Logger]
-  )(implicit trace: ZTraceElement): ZIO[Any, Throwable, Unit] = {
-    val argslayer: ULayer[ZIOAppArgs] =
-      ZLayer.succeed(
-        ZIOAppArgs(Chunk.empty)
+    spec: ZIOSpecAbstract
+  )(implicit trace: ZTraceElement): ZIO[Any, Throwable, Unit] =
+    (for {
+      _ <- ZIO.succeed("TODO pass this down through spec.runSpec to ExecutionEventSink: " + eventHandler)
```
Comment (adamgfraser) [inline test-sbt/shared/src/main/scala/zio/test/sbt/BaseTestTask.scala:56]: ???

File: test-tests/shared/src/test/scala/zio/test/ParallelSuitesInterleavedResultsSpec.scala [hunk 8]
```diff
@@ -0,0 +1,103 @@
+package zio.test
+
+import zio._
+
+object AMinimalSpec extends ZIOSpecDefault {
```
Comment (adamgfraser) [inline test-tests/shared/src/test/scala/zio/test/ParallelSuitesInterleavedResultsSpec.scala:5]: Do we want to keep this? I know this was useful for visual debugging during development but I don't think we are actually verifying any expected behavior in CI. Maybe we can have some flag for whether this runs so we can turn it back on when doing further development?

File: test-tests/shared/src/test/scala/zio/test/SpecSpec.scala [hunk 9]
```diff
@@ -187,5 +173,5 @@ object SpecSpec extends ZIOBaseSpec {
         }
       )
     }
-  )
+  ) @@ TestAspect.ignore // TODO Rewrite these to use final Summary instead of ExecutedSpec
```
Comment (adamgfraser) [inline test-tests/shared/src/test/scala/zio/test/SpecSpec.scala:176]: Can we minimize this any more? There are a lot of tests in here that are making assertions about behavior that is not related to the test rendering.

File: test/shared/src/main/scala/zio/test/DefaultTestReporter.scala [hunk 10]
```diff
@@ -26,136 +27,139 @@ import zio.{Cause, _}
...
+// TODO Needs to be re-written or simply dropped for new streaming behavior
 object DefaultTestReporter {
-  def render[E](
-    executedSpec: ExecutedSpec[E],
+  def apply[E](testRenderer: TestRenderer, testAnnotationRenderer: TestAnnotationRenderer)(implicit
+    trace: ZTraceElement
+  ): TestReporter[E] = { (duration: Duration, executedSpec: ExecutionEvent) =>
+    // val rendered = testRenderer.render(render(executedSpec, true), testAnnotationRenderer)
+    // val stats    = testRenderer.render(logStats(duration, executedSpec) :: Nil, testAnnotationRenderer)
+    val rendered = List.empty
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/DefaultTestReporter.scala:38]: Dead code?

**Review by adamgfraser — DISMISSED** [review]
Left a couple of suggestions for clean up but otherwise this looks great!

File: test-sbt/js/src/main/scala/zio/test/sbt/ZTestRunner.scala (hunk 3 above)
Comment (swoogles) [reply test-sbt/js/src/main/scala/zio/test/sbt/ZTestRunner.scala:115]: Updated comment to reference https://github.com/zio/zio/issues/6481

File: test-sbt/jvm/src/main/scala/zio/test/sbt/ZTestRunner.scala (hunk 2 above)
Comment (swoogles) [reply test-sbt/jvm/src/main/scala/zio/test/sbt/ZTestRunner.scala:153]: Nope, removed

File: test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala (hunk 6 above)
Comment (swoogles) [reply test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala:24]: Added new issue number here

File: test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala (hunk 5 above)
Comment (swoogles) [reply test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala:90]: Deleted

File: test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala (hunk 4 above)
Comment (swoogles) [reply test-sbt/jvm/src/test/scala/zio/test/sbt/ZTestFrameworkSpec.scala:96]: Do you mean this whole method? It's not being called at the moment because we switched from the SBT logger to the `Console`-based one, but I'm planning to restore this test once we've switched over to assertions on the `Console` output.

File: test-sbt/shared/src/main/scala/zio/test/sbt/BaseTestTask.scala (hunk 7 above)
Comment (swoogles) [reply test-sbt/shared/src/main/scala/zio/test/sbt/BaseTestTask.scala:56]: Did this as an alternative to a TODO comment + an arbitrary action on `eventHandler` so that it wasn't breaking the build as an unused parameter. I can remove it from this parameter list now, but I will probably need to add it back in for the `TestReport` follow-up work.
Let me know your preference

File: test-tests/shared/src/test/scala/zio/test/ParallelSuitesInterleavedResultsSpec.scala (hunk 8 above)
Comment (swoogles) [reply test-tests/shared/src/test/scala/zio/test/ParallelSuitesInterleavedResultsSpec.scala:5]: Should I just stick an `@@ ignore` on it for now?

File: test-tests/shared/src/test/scala/zio/test/SpecSpec.scala (hunk 9 above)
Comment (swoogles) [reply test-tests/shared/src/test/scala/zio/test/SpecSpec.scala:176]: Applied to only specific failing cases now.

File: test/shared/src/main/scala/zio/test/DefaultTestReporter.scala (hunk 10 above)
Comment (swoogles) [reply test/shared/src/main/scala/zio/test/DefaultTestReporter.scala:38]: Leaving this in for the moment, so that I can scrutinize it in isolation here.
https://github.com/zio/zio/issues/6484
Added ticket # to comments

**Review by adamgfraser — APPROVED** [review]
Congratulations! This is going to make viewing test output a lot nicer!
